Bluetooth: Controller: nrf54h20 cpurad support

cvinayak opened the pull request:
nrf54h20 bt_ll_sw_split support. Contributed/co-authored by @rick1082 

Building peripheral_hr for cpuapp:
```
west build --sysbuild -p -d build/peripheral_hr_cpuapp -b nrf54h20dk/nrf54h20/cpuapp samples/bluetooth/peripheral_hr
```

Flashing peripheral_hr to cpuapp:
```
west flash -d build/peripheral_hr_cpuapp
```

Building hci_ipc for cpurad:
```
west build --sysbuild -p -d build/hci_ipc_cpurad -b nrf54h20dk/nrf54h20/cpurad samples/bluetooth/hci_ipc
```

Flashing hci_ipc for cpurad:
```
west flash -d build/hci_ipc_cpurad
```


****

**Old before 20260107**:

**Broadcaster, non-connectable**:
```
west build --sysbuild -p -d build/broadcaster -b nrf54h20dk/nrf54h20/cpurad samples/bluetooth/broadcaster
```
```
west flash -d build/broadcaster
```

**Extended Scanning**:
```
west build --sysbuild -p -d build/observer -b nrf54h20dk/nrf54h20/cpurad samples/bluetooth/observer -DCONF_FILE=prj_extended.conf
```
```
west flash -d build/observer
```

Please have the below in an application core sample (`hello_world` sample):
`CONFIG_SOC_NRF54H20_CPURAD_ENABLE=y`

Local change to have `peripheral_hr` sample functional:
```
diff --git a/samples/bluetooth/peripheral_hr/prj.conf b/samples/bluetooth/peripheral_hr/prj.conf
index c24dc7c5967..bbbdcfcabc4 100644
--- a/samples/bluetooth/peripheral_hr/prj.conf
+++ b/samples/bluetooth/peripheral_hr/prj.conf
@@ -6,5 +6,7 @@ CONFIG_BT_DIS=y
 CONFIG_BT_DIS_PNP=n
 CONFIG_BT_BAS=y
 CONFIG_BT_HRS=y
-CONFIG_BT_DEVICE_NAME="Zephyr Heartrate Sensor"
+CONFIG_BT_DEVICE_NAME="Zephyr HRS nRF54H20"
 CONFIG_BT_DEVICE_APPEARANCE=833
+
+CONFIG_MAIN_STACK_SIZE=2048
diff --git a/samples/bluetooth/peripheral_hr/src/main.c b/samples/bluetooth/peripheral_hr/src/main.c
index 6a75f441f07..50de87a7b83 100644
--- a/samples/bluetooth/peripheral_hr/src/main.c
+++ b/samples/bluetooth/peripheral_hr/src/main.c
@@ -267,7 +267,7 @@ int main(void)
 
        /* Implement notification. */
        while (1) {
-               k_sleep(K_SECONDS(1));
+               k_yield();
 
                /* Heartrate measurements simulation */
                hrs_notify();
diff --git a/samples/hello_world/prj.conf b/samples/hello_world/prj.conf
index b2a4ba59104..c55b8a8c8a2 100644
--- a/samples/hello_world/prj.conf
+++ b/samples/hello_world/prj.conf
@@ -1 +1,5 @@
-# nothing here
+# Enable nrf54h20 cpurad
+CONFIG_SOC_NRF54H20_CPURAD_ENABLE=y
+
+CONFIG_MRAM_LATENCY=y
+CONFIG_MRAM_LATENCY_AUTO_REQ=y
```

cvinayak committed:
Bluetooth: Controller: nrf54h20 cpurad support

nrf54h20 cpurad bt_ll_sw_split support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: bap broadcast source/sink nRF54H20DK USB Audio

bap broadcast source/sink nRF54H20DK with USB Audio Device
support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Commit:
Bluetooth: Controller: nRF54H20 compile error fixes

FICR compile error fixed
ECB compile error fixed
SWI compile error fixed
draft modification for radio
Update radio.c due to miss some device check

Signed-off-by: Jui-Chou Chung <jui-chou.chung@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nrf54hx: Mitigate MRAM no latency control race

Added implementation to mitigate nrf54hx MRAM no latency
control race.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nrf54hx: Initial HF clock control

Added initial implementation for nrf54hx HF clock control.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nrf54hx: Initial GRTC support

Added initial implementation for nrf54hx GRTC use for Radio.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nrf54hx: Initial MRAM no latency control

Added initial implementation for nrf54hx MRAM no latency
control.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nrf54hx: Initial LF clock control

Added initial implementation for nrf54hx LF clock control.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: nRF54H20 cpurad hci_ipc quirks for ISO support

nRF54H20 cpurad hci_ipc quirks for ISO support

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: nRF54H20 cpurad hci_usb quirks for ISO support

nRF54H20 cpurad hci_usb quirks for ISO support

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: nRF54H20 cpuapp hci_usb quirks

nRF54H20 cpuapp hci_usb quirks.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: nRF54H20 cpuapp peripheral_hr quirks

nRF54H20 cpuapp peripheral_hr quirks.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54hx.h:
```diff
@@ -0,0 +1,817 @@
+{
+	/* TODO: Add any required cleanup of actions taken in hal_radio_reset()
+	 */
+	nrf_lrcconf_clock_always_run_force_set(NRF_LRCCONF000, 0, false);
```
Probably applies several places in this file (and maybe even most of the defines above?)

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi.h:
```diff
@@ -292,6 +292,14 @@ static inline void hal_trigger_aar_ppi_config(void)
 #if !defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER)
 
+/* Start SW-switch timer on event timer start.
+ */
```
```suggestion
/* Start SW-switch timer on event timer start. */
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ecb.c:
```diff
@@ -49,7 +56,7 @@ struct ecb_param {
+#if defined(NRF54L_SERIES) || defined(NRF54H_SERIES)
 	struct ecb_job_ptr in[2];
 	struct ecb_job_ptr out[2];
 #endif /* NRF54L_SERIES */
```
This is also missing several places further down in this file

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+	int err;
+
+	err = nrf_clock_control_release(clock_dev_lf, &clock_req_spec_lf);
+	if (err) {
```
```suggestion
	if (err != 0) {
```

Thalley commented on tests/bluetooth/controller/ctrl_periodic_sync/src/main.c:
```diff
@@ -346,26 +346,26 @@ ZTEST(periodic_sync_transfer, test_periodic_sync_transfer_rem_2)
 	struct pdu_data_llctrl_periodic_sync_ind local_periodic_sync_ind = {
 		.id = 0x00,
-		.conn_event_count = 0x01,
+		.conn_event_count = 0x0000,
```
```suggestion
		.conn_event_count = 0x0000U,
```

Thalley commented on tests/bluetooth/controller/ctrl_periodic_sync/src/main.c:
```diff
@@ -346,26 +346,26 @@ ZTEST(periodic_sync_transfer, test_periodic_sync_transfer_rem_2)
 		.phy = 0x01,
 		.adv_addr = { 0, 0, 0, 0, 0, 0},
-		.sync_conn_event_count = 0
+		.sync_conn_event_count = 0x0000,
```
```suggestion
		.sync_conn_event_count = 0x0000U,
```

Thalley reviewed (changes requested):
Overall the PR looks OK, and it's clear that a lot of it are based/copied from other existing files/code. Unfortunately this also means that there are a lot of coding style and guidelines violations that really should be applied to new code

cvinayak commented on tests/bluetooth/controller/ctrl_periodic_sync/src/main.c:
```diff
@@ -481,14 +481,14 @@ ZTEST(periodic_sync_transfer, test_periodic_sync_transfer_loc_twice)
 		.phy = 0x00,
 		.adv_addr = { 0, 0, 0, 0, 0, 0},
-		.sync_conn_event_count = 0
+		.sync_conn_event_count = 0xFFFF,
```
@Thalley Fixed the changes requested to this file/commit and move the commit to #101856, as that PR was failing. This PR will be rebase after #101856 is merged.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+static atomic_val_t hf_refcnt;
+
+#if defined(CONFIG_SOC_SERIES_NRF54HX)
+#define CLOCK_CONTROL_NRF_K32SRC_ACCURACY 7U /* FIXME: */
```
updated with a comment.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+#if defined(CONFIG_SOC_SERIES_NRF54HX)
+#define CLOCK_CONTROL_NRF_K32SRC_ACCURACY 7U /* FIXME: */
+
+const static struct device *clock_dev_lf = DEVICE_DT_GET(DT_NODELABEL(lfclk));
```
keeping as-is

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+
+#if !defined(CONFIG_BT_CTLR_ZLI)
+const static struct device *clock_dev_hfxo = DEVICE_DT_GET(DT_NODELABEL(hfxo));
+const static struct nrf_clock_spec clock_req_spec_hfxo = {
```
keeping as-is

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+	/* TODO: */
 
+	return 0;
+}
```
Updated comment.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -12,12 +12,44 @@
+{
+	return 0;
+}
+
```
stale comment... the function are available.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54hx.h:
```diff
@@ -0,0 +1,755 @@
+	HAL_RADIO_NS2US_ROUND(HAL_RADIO_NRF54HX_TXEN_TXIDLE_TX_S8_DEFAULT_NS)
+
+/* TXEN->TXIDLE + TXIDLE->TX (with default Radio ramp-up mode and
+ * no HW TIFS auto-switch) in microseconds for LE 2M PHY [S8].
```
done

cvinayak committed:
Bluetooth: Controller: nrf54h20 cpurad review rework

Review rework updates.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1360,7 +1367,8 @@ void radio_tmr_rx_status_reset(void)
-#if defined(CONFIG_SOC_COMPATIBLE_NRF5340_CPUNET) || defined(CONFIG_SOC_COMPATIBLE_NRF54LX)
+#if defined(CONFIG_SOC_COMPATIBLE_NRF5340_CPUNET) || defined(CONFIG_SOC_COMPATIBLE_NRF54LX) || \
+	defined(CONFIG_SOC_SERIES_NRF54HX)
 #else /* !CONFIG_SOC_COMPATIBLE_NRF5340_CPUNET && !CONFIG_SOC_COMPATIBLE_NRF54LX */
```
done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1380,15 +1389,17 @@ void radio_tmr_rx_enable(void)
+	defined(CONFIG_SOC_SERIES_NRF54HX)
 	nrf_radio_subscribe_clear(NRF_RADIO, NRF_RADIO_TASK_TXEN);
 #else /* !CONFIG_SOC_COMPATIBLE_NRF5340_CPUNET && !CONFIG_SOC_COMPATIBLE_NRF54LX */
 #endif /* !CONFIG_SOC_COMPATIBLE_NRF5340_CPUNET && !CONFIG_SOC_COMPATIBLE_NRF54LX */
```
done

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1876,9 +1890,12 @@ void radio_tmr_end_capture(void)
+	  defined(CONFIG_SOC_SERIES_NRF54HX)) && \
 	 !defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER))
 	hal_radio_end_time_capture_ppi_config();
 	hal_radio_nrf_ppi_channels_enable(BIT(HAL_RADIO_END_TIME_CAPTURE_PPI));
```
done

Thalley commented on samples/bluetooth/bap_broadcast_source/boards/nrf54h20dk_nrf54h20_cpuapp.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_UDC_DRIVER_HIGH_SPEED_SUPPORT_ENABLED=n
+
+# Enable Power Management for nRF54H20
+CONFIG_PM=y
```
Ditto as with the sink sample comments

Thalley commented on samples/bluetooth/bap_broadcast_source/boards/nrf54h20dk_nrf54h20_cpuapp.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_SAMPLE_USBD_PRODUCT="USB Broadcast Audio nRF54H20DK"
+
+# Default to 24 KHz sampling for nRF54H20 until we figure why 48KHz does not work efficiently
+CONFIG_BAP_BROADCAST_24_2_1=y
```
Why not just use the sample default instead?

Thalley commented on samples/bluetooth/hci_ipc/extra-iso-bt_ll_sw_split.conf:
```diff
@@ -163,6 +169,8 @@ CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL=y
-CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG=y
+#
+# Upstream nRF54H20 has no true random number generator driver, lets keep this disabled
+# CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG=y
```
This disables `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` for all platforms, and not just the nRF54H20. Suggest to move this change to a board specific file

Thalley commented on samples/bluetooth/hci_ipc/extra-iso-bt_ll_sw_split.conf:
The rename does not match the file anymore IMO. 

Previously it was an _alternative_ conf file, but now you've prefixed it with `extra` implying that it should now be used as an `EXTRA_CONF_FILE`, right? 

If that is the case, then anything in `prj.conf`  that is enabled like `CONFIG_BT` and `CONFIG_IPC_SERVICE` that remain unchanged should be removed from this file

Thalley commented on samples/bluetooth/hci_usb/boards/nrf54h20dk_nrf54h20_cpuapp.conf:
```diff
@@ -0,0 +1,11 @@
+CONFIG_SOC_NRF54H20_CPURAD_ENABLE=y
+
+# We are only using full-speed, disable including high-speed implementation
+CONFIG_UDC_DRIVER_HIGH_SPEED_SUPPORT_ENABLED=n
```
If the sample only supports full-speed, then shouldn't this be applied to all platforms? If it supports high-speed, shouldn't this just use the default value?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf54hx.h:
```diff
@@ -0,0 +1,817 @@
+{
+	/* TODO: Add any required cleanup of actions taken in hal_radio_reset()
+	 */
+	nrf_lrcconf_clock_always_run_force_set(NRF_LRCCONF000, 0, false);
```
```suggestion
	nrf_lrcconf_clock_always_run_force_set(NRF_LRCCONF000, 0U, false);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+const static struct nrf_clock_spec clock_req_spec_lf = {
+	.frequency = 32768,
+	.accuracy = sca_ppm_lut[CLOCK_CONTROL_NRF_K32SRC_ACCURACY],
+	.precision = 0, /* 0 for low precision, 1 for high precision */
```
Suggest to use defines for the frequency and accuracy, or at least provide references to where these very specific values are defined

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+	sys_notify_init_spinwait(&clock_cli_lf.notify);
 
+	err = nrf_clock_control_request(clock_dev_lf, &clock_req_spec_lf, &clock_cli_lf);
+	if (err) {
```
```suggestion
	if (err != 0) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+const static struct nrf_clock_spec clock_req_spec_hfxo = {
+	.frequency = 0, /* Ignore or use default */
+	.accuracy = 1, /* Use maximum accuracy */
+	.precision = 1, /* 0 for low precision, 1 for high precision */
```
```suggestion
	.frequency = 0U, /* Ignore or use default */
	.accuracy = 1U, /* Use maximum accuracy */
	.precision = 1U, /* 0 for low precision, 1 for high precision */
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c:
```diff
@@ -10,21 +10,128 @@
+	int err;
+
+	err = nrf_clock_control_release(clock_dev_hfxo, &clock_req_spec_hfxo);
+	if (err) {
```
```suggestion
	if (err != 0) {
```

Thalley reviewed (changes requested):
I think this PR is doing _entirely_ too many things, and probably can/should be split into 5-10 other PRs. 

Additionally it also contains a "fixup" commit which it shouldn't

cvinayak committed:
Bluetooth: Controller: nrf54hx: Encryption and Privacy support

Add use of NRF_ECB, NRF_CCM and NRF_AAR to have encryption
and accelerated address resolution support in nRF54H20 SoC.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mavik1 commented on samples/bluetooth/bap_broadcast_sink/sysbuild.cmake:
```diff
@@ -13,8 +13,8 @@ if(SB_CONFIG_NET_CORE_IMAGE_HCI_IPC)
     BOARD       ${SB_CONFIG_NET_CORE_BOARD}
```
`Kconfig.sysbuild` should be updated to ensure that `SB_CONFIG_NET_CORE_BOARD` defaults to "nrf54h20dk/nrf54h20/cpurad" when targeting the nrf54h20dk.

mavik1 commented on samples/bluetooth/bap_broadcast_source/sysbuild.cmake:
```diff
@@ -13,8 +13,8 @@ if(SB_CONFIG_NET_CORE_IMAGE_HCI_IPC)
     BOARD       ${SB_CONFIG_NET_CORE_BOARD}
```
`Kconfig.sysbuild` should be updated to ensure that `SB_CONFIG_NET_CORE_BOARD` defaults to "nrf54h20dk/nrf54h20/cpurad" when targeting the nrf54h20dk.

cvinayak commented:
Rebased to latest upstream  main. Added updates to broadcaster sample.

Broadcaster with BT Enable and Disable.... does not get enabled after being disable.. pending investigation.

```
west build --sysbuild -p -d build/broadcaster_nrf54h20dk_cpuapp -b nrf54h20dk/nrf54h20/cpuapp samples/bluetooth/broadcaster -- -DCONFIG_SAMPLE_BT_USE_DISABLE=y -DCONFIG_SERIAL=n -DCONFIG_GPIO=n
```

```
west flash -d build/broadcaster_nrf54h20dk_cpuapp
```

```
west build --sysbuild -p -d build/hci_ipc_nrf54h20dk_cpurad -b nrf54h20dk/nrf54h20/cpurad samples/bluetooth/hci_ipc -- -DCONFIG_SERIAL=n -DCONFIG_GPIO=n
```

```
west flash -d build/hci_ipc_nrf54h20dk_cpurad
```


Power profile, between events, ~7.39 uA CPU system-on-idle:
<img width="2858" height="1680" alt="image" src="https://github.com/user-attachments/assets/845ca4d8-4073-4e2e-a7e9-59a4976e4e09" />

Power profile, after bt_disable, ~8 uA CPU system-on-idle:
<img width="2858" height="1680" alt="image" src="https://github.com/user-attachments/assets/6c3ff02c-dbdc-4289-8112-ae21bab95315" />
